### ISSUE TITLE
Any chance to integrate OmniSharp?

### ISSUE THREAD

**Issue by partounian** [w0rp/ale#960]
https://github.com/OmniSharp/omnisharp-vim

**Comment by w0rp** [comment]
I'm not sure. I don't know how it works. Someone else might.

**Comment by stuart-bennett** [comment]
Caveat - I'm new to extending Vim + Ale so my observations might not be totally accurate or completely unfounded :)

OmniSharp doesn't fit as neatly into the typical usage I've seen of Ale. By that I mean typically there's a standalone command that performs the linting whereas OmniSharp runs as a separate process (using HTTP to handle back and forth between vim), it loads project files at startup to bootstrap a lot of rich functionality outside of linting.

It does however have a /codecheck endpoint that does lint source code and the OmniSharp-Vim plugin exposes this via Omnisharp#CodeCheck() function.

@w0rp - is it possible / can you suggest anyways to define a linter like (I know return types aren't correct but this how I could envisage it working):

```
function! ale_linters#cs#GetCommand(buffer) abort
    	" From omnisharp-vim, returns 0 || 1
	return g:OmniSharp_loaded
endfunction

function! ale_linters#cs#Executable(buffer) abort
	call OmniSharp_CodeCheck()
endfunction

call ale#linter#Define('cs', {
\   'name': 'omnisharp',
\   'executable_callback': ale_linters#cs#Executable
\   'command_callback': ale_linters#cs#GetCommand
...
```

So basically instead of calling out to an external cmd, invoke a Vim function? Feels like this might be outside the use-case for Ale, and should perhaps just write a custom function to invoke the Vim function and manually add the results to the linting errors collection?

**Comment by w0rp** [comment]
I'm not sure, that sounds a bit unusual.

**Comment by stuart-bennett** [comment]
Yeah, I suppose fundamentally it operates like Language Specification Protocol Server would, in fact the OmniSharp team appear to be working on that (https://github.com/OmniSharp/csharp-language-server-protocol)
I didn't notice that Ale had support for LSP so maybe this will be a lot easier when this OmniSharp LSP implementation is available.

**Comment by w0rp** [comment]
Integrating with an LSP server is a better bet.

**Comment by jmg5e** [comment]
You should maybe check out (https://github.com/OmniSharp/omnisharp-node-client). I have had success getting ominsharp support using the [vim-lsp plugin](https://github.com/prabirshrestha/vim-lsp). What all goes into defining linters that take advantage of ale's lsp capabilities?

**Comment by w0rp** [comment]
Have a look at the `rls` or `pyls` linters for examples. There are a few other examples. Also consult the documentation for `ale#linter#Define`.

**Comment by JDeuce** [comment]
Ah damn I just switched from syntactic to ale recently and I want to start using omnisharp. The integration with synthetic is working fine so I guess the solution is to just go back to using that for now.

**Comment by w0rp** [comment]
If you really want it, you could open a pull request for it.

**Comment by JulioJu** [comment]
Actually, OmniSharp works well with ALE. Maybe https://github.com/w0rp/ale/blob/master/doc/ale-cs.txt and https://github.com/w0rp/ale#1-supported-languages-and-tools could be updated ? And this current Issue could be closed ?

**Comment by w0rp** [comment]
I'll add some text in the documentation. Thanks for the information!
